File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        lastSeen = {}
        for i in range(len(nums)):
            lastSeen[nums[i]] = i

        solutions = set()
        for i in range(len(nums) - 2):
            for j in range(i + 1, len(nums) - 1):
                needed = -(nums[i] + nums[j])
                if needed in lastSeen and lastSeen[needed] > j:
                    solutions.add((nums[i], nums[j], needed))
        
        return solutions
        

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        set<tuple<int, int, int>> found;
        vector<vector<int>> results;
    
        // Old python solution simplified to 2-sum n times
        // Was somehow too slow, looking into hint "2 pointer" instead of hash
        for (int i = 0; i < nums.size() - 2; i++){
            // Idea: fix num1 to be at i
            // Need to get other 2 = -num1. Start at smallest + biggest
            // Shift smallest and biggest when sum > or < 0

            int target = -nums[i];
            int small = i+1;
            int big = nums.size() - 1;

            while (small < big){
                int total = nums[small] + nums[big];

                if (total == target){
                    // add solution
                    tuple<int, int, int> seen = make_tuple(nums[i], nums[small], nums[big]);
                    if (!found.count(seen)){
                        found.insert(seen);
                        vector<int> sol = {nums[i], nums[small], nums[big]}; 
                        results.push_back(sol);
                    }
                    small++;
                    big--;

                }else if (total < target){
                    small++;
                }else{
                    big--;
                }

            }
        }

        return results;
    }
};
